File: src/383. Ransom Note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */

export function canConstruct(ransomNote, magazine) {
  let hash = {}
  for (let i = 0; i < magazine.length; i++) {
    let alphabet = magazine[i]
    if (!hash[alphabet]) hash[alphabet] = 1
    else hash[alphabet]++
  }
  for (let j = 0; j < ransomNote.length; j++) {
    let alphabet = ransomNote[j]
    if (hash[alphabet] === 0 || hash[alphabet] === undefined) return false
    else if (hash[alphabet] > 0) hash[alphabet]--
  }
  return true
}

File: src/383. Ransom Note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> counter;
        for (auto &word : magazine) {
            counter[word]++;
        }
        for (auto &word : ransomNote) {
            if (counter.find(word) == counter.end() || counter[word] <= 0) return false;
            counter[word]--;
        }
        return true;
    }
};